### ISSUE TITLE
Data API filter refactor, logicalExpression

### ISSUE THREAD

**Issue by Yuqi-Du** [stargate/data-api#1389]
Improve Encapsulation.
Currently we have this recursive defined data structure LogicalExpression, basically because we need to keep track of the AND/OR relation of filter entries.  And, we don’t want to lose these AND/OR information, so we just modify this LogicalExpression in-place, and pass down to operation level, and construct CQL query from it.  

However, this is breaking encapsulation since DB operation should not know LogicalExpression which is in the API level.

In order to improve the encapsulation, we need to refactor the FilterMatcher, FilterMatcherRule, FilterMatcherRules, FilterResolver..etc. 
Basically, we need similar data structure to represent logical relation, and it will be capture and converted when we resolve API filters to DB filters.
